### ISSUE TITLE
Regarding Hyperparameters

### ISSUE THREAD

**Issue by sriharsha0806** [tonylins/pytorch-mobilenet-v2#1]
Thanks for opensourcing the code so quick.
can you let me know the Hyperparameters you used? I want to train my own?  Did you set batchsize 96 only?  
and 
net = models.MobileNetV2(n_class=1000)
I am getting error "name 'models' is not defined". 

Did you understand the Theorm 1?

**Comment by tonylins** [comment]
Hi,

Thanks for finding out the error. It's a typo and I've fixed it with an update.

The Hyper-parameters are the same as in the paper, except that I used batch size 256, which could lead to potential performance degradation.

I didn't have time to read the supplementary by now.

**Comment by sriharsha0806** [comment]
Hi, are you training on Multiple GPUs ?

**Comment by tonylins** [comment]
Yes, on 4 GPUs.

**Comment by ildoonet** [comment]
@tonylins so, if you trained it with 4 gpus, you split 256 batch to 256/4=64 batches for each gpu tower. Am I right?

**Comment by tonylins** [comment]
@ildoonet Yes

**Comment by lcwyylcwyy** [comment]
How many epochs in your training?

**Comment by lcwyylcwyy** [comment]
How many epochs in your training? @tonylins

**Comment by hetong007** [comment]
Can you please let us know that have you:
- changed rmsprop to sgd (or nesterov)?
- followed the paper that starting with learning rate=0.045 and decay by 0.98 for every epoch?

Thanks!

**Comment by mathmanu** [comment]
Could you please share the training script if possible?

**Comment by blueardour** [comment]
@tonylins   Hi, Thanks for the effort in sharing the mobilenet-v2. I followed your network structure and trained it with SGD, however it could only reach 66% top1 and 88% top5. Big gap from the 71.8% top1.

In my train, the batch_size was set 128, lr began with 0.045 and a 0.1 drop after plateau detection strategy was employed. I wondered if you could share some of your experience and give me a light to gain better result. Really appreciate it.

**Comment by tonylins** [comment]
Hi @blueardour, 
I used batch_size 256 and init lr=0.045, decayed by 0.98 every epoch, and trained for 240~300 epochs. Also, it is recommended to use drop out rate 0.2 for faster training.

The data processing is:

normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                         std=[0.229, 0.224, 0.225])

        input_size = 224
        train_dataset = datasets.ImageFolder(
            traindir,
            transforms.Compose([
                transforms.RandomResizedCrop(input_size, scale=(0.2, 1.0)),  # TODO: find a good scale ratio
                transforms.RandomHorizontalFlip(),
                transforms.ToTensor(),
                normalize,
            ]))

        train_loader = torch.utils.data.DataLoader(
            train_dataset, batch_size=batch_size, shuffle=(train_sampler is None),
            num_workers=n_worker, pin_memory=True, sampler=train_sampler)

        val_loader = torch.utils.data.DataLoader(
            datasets.ImageFolder(valdir, transforms.Compose([
                transforms.Resize(int(input_size/0.875)),
                transforms.CenterCrop(input_size),
                transforms.ToTensor(),
                normalize,
            ])),

**Comment by blueardour** [comment]
Hi, tonylins. Thanks for your detail description.
Compared with your data processing, my experiment was different with：
1. the augmentation on the train set. I didn't employ the RandomResizedCrop but only a rand crop after resize original image to 256*256.
2. Also, I only train for 90 epochs.
3. I didn't apply the dropout.
Maybe these are the reasons for the accuracy gap, specially the data augment one.

My final train result:  top1 and top5 on training set: 73.1/ 91.84; top1 and top5 on evalution set: 67/ 88.17. It could be found that some bit of over-fitting in the training. Maybe the dropout can alleviate this.
By the way, where did you add the dropout? The input layer or certain deep layer?

**Comment by cxxgtxy** [comment]
Hi,
I use the hyper-parameters as your suggestion. After I trained 300 epochs, the top 1 is 71.41%. Any ideas about this? I check the official TF implementation, the smooth loss is applied. I wonder whether you use smooth loss or not. Thanks.

**Comment by hetong007** [comment]
@cxxgtxy In GluonCV we get 71.92%: https://gluon-cv.mxnet.io/model_zoo/classification.html#mobilenet

Our training parameters is here: https://raw.githubusercontent.com/dmlc/web-data/master/gluoncv/logs/classification/imagenet/mobilenetv2_1.0.sh
Our training log is here: https://raw.githubusercontent.com/dmlc/web-data/master/gluoncv/logs/classification/imagenet/mobilenetv2_1.0.log

PyTorch should have similar results with similar parameters.

**Comment by cxxgtxy** [comment]
@hetong007  Thanks!

**Comment by D-X-Y** [comment]
@hetong007 Hi, thanks for your reference. DId you GluonCV use smooth loss or not?

**Comment by tonylins** [comment]
> Hi,
> I use the hyper-parameters as your suggestion. After I trained 300 epochs, the top 1 is 71.41%. Any ideas about this? I check the official TF implementation, the smooth loss is applied. I wonder whether you use smooth loss or not. Thanks.

Hi, recently I found a more efficient setting by training only for 150epochs using cosine learning rate starting from 0.05 with weight decay 4e-5. No label smoothing is needed and you should get around 72% accuracy. 

Also use dropout=0 for training.

**Comment by D-X-Y** [comment]
@tonylins Cool, I will have a try!

**Comment by tonylins** [comment]
@D-X-Y Be sure to use dropout=0, which I just added.

**Comment by D-X-Y** [comment]
@tonylins ok, thanks for your suggestion.

**Comment by cxxgtxy** [comment]
@tonylins COOL, I will try.

**Comment by mathmanu** [comment]
@tonylins Could you please clarify what you mean by "Be sure to use dropout=0, which I just added".

I can see that you are using nn.Dropout(0.2) in the code. (https://github.com/tonylins/pytorch-mobilenet-v2/blob/master/MobileNetV2.py)

**Comment by D-X-Y** [comment]
@tonylins May I ask the total batch size, and the number of GPUs that you use?

**Comment by tonylins** [comment]
@D-X-Y used batch size 256, 2 or 4 GPUs.

**Comment by D-X-Y** [comment]
@tonylins Following your suggestion, I have trained 46 epochs, and the validation accuracy is 62.44. Is that similar to your training log? In addition, would you mind to share your training log with me for references?

**Comment by D-X-Y** [comment]
Hi guys, I obtain 72% accuracy by using the suggested hyper-parameters from @tonylins. Thanks a lot for all the above discussion.

**Comment by cxxgtxy** [comment]
@D-X-Y Your training speed is so high. My training speed is about 40mins per epoch.
And I will report my results when the training ends.

**Comment by cxxgtxy** [comment]
@tonylins  I use your latetest recommended hyper-parameters(cos 150 epcoch), after trained 120epochs, the validation acc is about  69.75. Can you check your training log? Thanks. 
thanks@D-X-Y

**Comment by cxxgtxy** [comment]
@tonylins I also obtain a good top 1, which is more than 72%.
Thanks

**Comment by tonylins** [comment]
@cxxgtxy @D-X-Y Hi guys, I have added the training recipe to README. I will close the issue for now. Feel free to reopen it if you encounter other problems!

**Comment by wangdongxuking61** [comment]
> Hi guys, I obtain 72% accuracy by using the suggested hyper-parameters from @tonylins. Thanks a lot for all the above discussion.

could you please share your training script? Thanks a lot. I can only get 71.3%.
BTW, did you use warmup?

**Comment by D-X-Y** [comment]
@wangdongxuking61 I shared the script to train MobileNet here: https://github.com/D-X-Y/NAS-Projects/blob/master/BASELINE.md#performance-on-imagenet

**Comment by wangdongxuking61** [comment]
> @wangdongxuking61 I shared the script to train MobileNet here: https://github.com/D-X-Y/NAS-Projects/blob/master/BASELINE.md#performance-on-imagenet

3Q very much! ^_^

**Comment by blueardour** [comment]
Hi, 

For who concerns, I got 73.164% Top-1 accuracy for the MBv1. The key hyper-parameter is to set the initial learning rate to 1e-1.

```
options="$options --lr 1e-1 --lr_policy sgdr --lr_custom_step 120 --wd 1e-4 --nesterov
```

**Comment by mathmanu** [comment]
@blueardour Thank you for sharing this. Can you also post the training transforms? How many total epochs and what is the cycle length in sgdr? If you can point to the training repository/code that will be even better.

**Comment by blueardour** [comment]
@mathmanu  Hi, I have the same model definition with this repo. So by using the dedicated hyerparamer is expected to obtain the 73% performance by the author's code.

For your question, I trained 120 epochs with SGDR without restart .  If you'd like to make clear of every detail of my training, refer my repo https://github.com/blueardour/model-quantization. Above accuracy is achieved with  config file: config/config.baseline.train-strach.imagenet.fp.mobilenet-v1  (This repo is oriented for model quantization of different tasks and is not fully released. Thus, you might find many extra logic and even uncomplete ones. Just neglect those if you are focused only in full precision mobilenet training.)

ps. For parts of my training history:

Paper | Dataset | Network | Bit(A/W/G) | Paper report | My Accuracy | Comment
----- | ---- | ---------- | ------ | -------  |  ---------
 -    | imagenet | v1 | 32/32/32 | 70.6 | 69.204 | cbas-0,preBN,fp16,wd5e-5,aug
 -    | imagenet | v1 | 32/32/32 | 70.6 | 69.800 | cbas-1,wd1e-4,fp32,aug
 -    | imagenet | v1 | 32/32/32 | 70.6 | 69.178 | cbas-2,wd1e-4,fp32
 -    | dali     | v1 | 32/32/32 | 70.6 | 65.568 | cbas-3,
 -    | imagenet | v1 | 32/32/32 | 70.6 | 72.446 | cbas-4,bs=256,fp32,wd1e-4
 **-    | imagenet | v1 | 32/32/32 | 70.6 | 73.164 | cbas-5**
 -    | imagenet | v1 | 32/32/32 | 70.6 | 55.780 | cbas-5, group-norm=8
 -    | imagenet | v1 | 32/32/32 | 70.6 | 58.246 | cbas-5, group-norm=16
 -    | imagenet | v1 | 32/32/32 | 70.6 | 70.310 | bacs-1,preBN
 -    | imagenet | v1 | 32/32/32 | 70.6 | 67.200 | bacs-3,preBN
 -    | imagenet | v1 | 32/32/32 | 70.6 | 71.652 | bacs-4,preBN
 -    | imagenet | v2 | 32/32/32 | - | 70.416 | cbas-1,relu6
 -    | imagenet | v2 | 32/32/32 | - | 72.288 | cbas-4
 -    | imagenet | v2 | 32/32/32 | - | 71.800 | bacs-5
